README.md
# WebSweeper
 A minesweeper game for the web. Completed for COM S 319 class at ISU.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Types of changes:
* `Added` for new features.
* `Changed` for changes in existing functionality.
* `Deprecated` for soon-to-be removed features.
* `Removed` for now removed features.
* `Fixed` for any bug fixes.
* `Security` in case of vulnerabilities.

## [Unreleased]

## [0.2.0] - 2019-04-12
#### Added
* Changelog file
* Timer for score of the game




server/test/test_server.js
const chai  = require('chai');
const chaiHttp  = require('chai-http');
const server = require('../server');
const should = chai.should();

chai.use(chaiHttp);

it('GET connect API', (done) => {
    chai.request(server)
    .get('/api/connect')
    .end((err, res) => {
        res.should.have.status(200);
        done();
    });
});

describe('GET new game API', () => {
    it('GET new game difficulty \'Easy\'', (done) => {
        chai.request(server)
        .get('/api/newgame/easy')
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.be.a('object');
            res.body.should.have.property('row').eql(5);
            res.body.should.have.property('col').eql(5);
            res.body.should.have.property('field');
            done();
        });
    });

    it('GET new game difficulty \'Medium\'', (done) => {
        chai.request(server)
        .get('/api/newgame/medium')
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.be.a('object');
            res.body.should.have.property('row').eql(10);
            res.body.should.have.property('col').eql(10);
            res.body.should.have.property('field');
            done();
        });
    });

    it('GET new game difficulty \'Hard\'', (done) => {
        chai.request(server)
        .get('/api/newgame/hard')
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.be.a('object');
            res.body.should.have.property('row').eql(15);
            res.body.should.have.property('col').eql(15);
            res.body.should.have.property('field');
            done();
        });
    });
});

describe('GET new game API', () => {
    it('GET leaderboard for difficulty \'Easy\'', (done) => {
        chai.request(server)
        .get('/api/leaderboard/easy')
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.be.a('object');
            res.body.should.have.property('list');
            res.body.list.should.have.have.lengthOf(10);
            done();
        });
    });

    it('GET leaderboard for difficulty \'Medium\'', (done) => {
        chai.request(server)
        .get('/api/leaderboard/medium')
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.be.a('object');
            res.body.should.have.property('list');
            res.body.list.should.have.have.lengthOf(10);
            done();
        });
    });

    it('GET leaderboard for difficulty \'Hard\'', (done) => {
        chai.request(server)
        .get('/api/leaderboard/hard')
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.be.a('object');
            res.body.should.have.property('list');
            res.body.list.should.have.have.lengthOf(10);
            done();
        });
    });
});

describe('POST check score API', () => {
    it('POST check score \'Easy\'', (done) => {
        let score = {
            "difficulty" : "easy",
            "score" : 300
        };

        chai.request(server)
        .post('/api/score')
        .send(score)
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.have.property('result');
            done();
        });
    });

    it('POST check score \'Medium\'', (done) => {
        let score = {
            "difficulty" : "medium",
            "score" : 400
        };

        chai.request(server)
        .post('/api/score')
        .send(score)
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.have.property('result');
            done();
        });
    });

    it('POST check score \'Hard\'', (done) => {
        let score = {
            "difficulty" : "hard",
            "score" : 500
        };

        chai.request(server)
        .post('/api/score')
        .send(score)
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.have.property('result');
            done();
        });
    });
});

describe('POST insert new score API', () => {
    it('POST insert score \'Easy\'', (done) => {
        let score = {
            "difficulty" : "easy",
            "score" : 15000,
            "username" : "testuser"
        };

        chai.request(server)
        .post('/api/score')
        .send(score)
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.have.property('result');
            res.body.should.have.property('list');
            res.body.list.should.have.have.lengthOf(10);
            done();
        });
    });

    it('POST insert score \'Medium\'', (done) => {
        let score = {
            "difficulty" : "medium",
            "score" : 15000,
            "username" : "testuser"
        };

        chai.request(server)
        .post('/api/score')
        .send(score)
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.have.property('result');
            res.body.should.have.property('list');
            res.body.list.should.have.have.lengthOf(10);
            done();
        });
    });

    it('POST insert score \'Hard\'', (done) => {
        let score = {
            "difficulty" : "hard",
            "score" : 15000,
            "username" : "testuser"
        };

        chai.request(server)
        .post('/api/score')
        .send(score)
        .end((err, res) => {
            res.should.have.status(200);
            res.body.should.have.property('result');
            res.body.should.have.property('list');
            res.body.list.should.have.have.lengthOf(10);
            done();
        });
    });
});

src/index.js
import React from 'react';
import './index.css';
import ReactDOM from 'react-dom';
import Cell from './Cell';
import ServerApi from './ServerApi'; // import the server api class

let array = [0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,1,0,1,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,1]
let x = 0; 
var squares;
var mineTime = 0;
let enteredUserName = "";
var leaderboard;

let startTime = new Date().getTime();
let endTime = 0;

let server = new ServerApi(); // server object

// prompt user for difficulty and store value as a variable
let difficulty = Number(prompt("Enter a difficulty (1-3)"));

// send the difficulty to the server so it can render the proper sized board
function setDifficulty(difficulty) {
  let difficultyString = ""; // The numbers converted to a string value
  switch (difficulty) {
    case 1:
      difficultyString = "easy";
      break;
    case 2:
      difficultyString = "medium";
      break;
    case 3:
      difficultyString = "hard";
      break;
    default:
      difficultyString = 'easy'; // a valid difficulty has not been entered
      break;
  }
  return difficultyString;
}

function test()
{
	alert("bleh");
}



export default class Board extends React.Component {
  state = {
    boardData: this.initBoardData(this.props.height, this.props.width, this.props.mines),
    gameStatus: "GalaxySweeper"
  };
  
  


  // Gets initial board data
  initBoardData(height, width, mines) {
    let data = this.createEmptyArray(height, width);
	data = this.plantMines(data, height, width, mines);
	let updatedData = data;
	for (let i = 0; i < height; i++) 
	{
      for (let j = 0; j < width; j++)
	  {
        if (data[i][j].isMine !== true) 
		{
          let mine = 0;
		  let adjSquares = this.traverseBoard(data[i][j].x, data[i][j].y, data);
		  adjSquares.map(value => {
            if (value.isMine) {
              mine++;
            }
          });
		  if (mine === 0) 
		  {
            updatedData[i][j].isEmpty = true;
          }
		  updatedData[i][j].neighbour = mine;
		}
	  }
	}
	data = updatedData;
    return data;
  }

  createEmptyArray(height, width) {
    let data = [];

    for (let i = 0; i < height; i++) {
      data.push([]);
      for (let j = 0; j < width; j++) {
        data[i][j] = {
          x: i,
          y: j,
          neighbour: 0,
        };
      }
    }
    return data;
  }
  
  plantMines(data, height, width, mines) {
    let randomx, randomy, minesPlanted = 0;

    while (minesPlanted < mines) {
      randomx = Math.floor(Math.random() * width);
      randomy = Math.floor(Math.random() * height);
      if (!(data[randomx][randomy].isMine)) {
        data[randomx][randomy].isMine = true;
        minesPlanted++;
      }
    }

    return (data);
  }

 
	
  renderBoard(data) {
    return data.map((datarow) => {
      return datarow.map((dataitem) => {
        return (
          <div key={dataitem.x * datarow.length + dataitem.y}>
            <div> <Cell onClick={() => this.handleCellClick(dataitem.x, dataitem.y)} value={dataitem}/></div>
            {(datarow[datarow.length - 1] === dataitem) ? <div className="clear" /> : ""}
          </div>);
      })
    });

  }
  
  traverseBoard(x, y, data) {
    const el = [];

    //up
    if (x > 0) {
      el.push(data[x - 1][y]);
    }

    //down
    if (x < this.props.height - 1) {
      el.push(data[x + 1][y]);
    }

    //left
    if (y > 0) {
      el.push(data[x][y - 1]);
    }

    //right
    if (y < this.props.width - 1) {
      el.push(data[x][y + 1]);
    }

    // top left
    if (x > 0 && y > 0) {
      el.push(data[x - 1][y - 1]);
    }

    // top right
    if (x > 0 && y < this.props.width - 1) {
      el.push(data[x - 1][y + 1]);
    }

    // bottom right
    if (x < this.props.height - 1 && y < this.props.width - 1) {
      el.push(data[x + 1][y + 1]);
    }

    // bottom left
    if (x < this.props.height - 1 && y > 0) {
      el.push(data[x + 1][y - 1]);
    }

    return el;
  }
  
  handleCellClick(x, y) {

    if (this.state.boardData[x][y].isRevealed || this.state.boardData[x][y].isFlagged) return null;
    let updatedData = this.state.boardData;
    updatedData[x][y].isRevealed = true;
	squares = squares - 1;
	console.log(squares);

	if(updatedData[x][y].isEmpty){
		updatedData = this.revealEmpty(x, y, updatedData);
	}
	
	if (this.state.boardData[x][y].isMine) {
      this.setState({ gameStatus: "You Lost." });
	  updatedData = this.revealAll(x, y, updatedData);
    }
	
//	if (updatedData[x][y].isEmpty) {
 //     updatedData = this.revealEmpty(x, y, updatedData);
//  }
	  
	else if (squares == this.props.mines)
	{
    this.setState({ gameStatus: "You Win." });
    endTime = new Date().getTime() - startTime;
    let gameScore = endTime / 1000;
    updatedData = this.revealAllWin(x, y, updatedData);
    enteredUserName = prompt("Congratulations on your win! Please enter your name. Score:" + gameScore); // prompt the user for their name
    server.insertNewScore(setDifficulty(difficulty), gameScore, enteredUserName); // TODO send the data to the server (score must be updated)
	}




    this.setState({
      boardData: updatedData,
    });
  }

	revealEmpty(x, y, data) {
	
    let area = this.traverseBoard(x, y, data);
    area.map(value => {
      if (!value.isRevealed && (value.isEmpty || !value.isMine)) {
        data[value.x][value.y].isRevealed = true;
		squares = squares - 1;
		console.log(squares);
        if (value.isEmpty) {
          this.revealEmpty(value.x, value.y, data);
		  
        }
      }
    });
    return data;
	

  }
  
  revealAll(x, y, data) {
    let area = this.state.boardData
    area.map((datarow) => {
		datarow.map((dataitem) => {
			dataitem.isRevealed = true;
		});
	});
	area.map((datarow) => {
		datarow.map((dataitem) => {
			dataitem.isRunning = false;
		});
	});
    return area;
}

  revealAllWin(x, y, data) {
    let area = this.state.boardData
    area.map((datarow) => {
		datarow.map((dataitem) => {
			dataitem.isRevealed = true;
		});
	});
	area.map((datarow) => {
		datarow.map((dataitem) => {
			dataitem.isWin = true;
		});
	});
    return area;
}
  
  render() {
	if (array[x] === 1)
	{
    return (
      <div className="board">
        <div className="game-info">
          <h1 className="info">{this.state.gameStatus}</h1>
        </div>
        {
          this.renderBoard(this.state.boardData)
		  
        }
      </div>
    );
	}
	if (array[x] === 0)
	{
    return (
      <div className="board">
        <div className="game-info">
          <h1 className="info">{this.state.gameStatus}</h1>
        </div>
        {
          this.renderBoard(this.state.boardData)
        }
      </div>
	  
    );
	}
  }
}

class Game extends React.Component {
	
  state = {
    className: "game"
  };
  
  componentDidMount() {
    // get new game from server and 
    //   initialize state values for the component
    let server = new ServerApi();
    server.getNewGame(setDifficulty(difficulty))
    .then(res => {
      this.setState({
        height: res.row,
        width : res.col,
        mines: res.mine,
        className: res.gameType,
      });

      array = [];
      for (let i = 0; i < res.field.length; i++) {
        array[i] = Number(res.field[i]);
      }
    squares = this.state.height * this.state.width;
    server.getLeaderboard(setDifficulty(difficulty))
	  .then(res => {
		  console.log(res);
		  document.getElementById("Leaderboard").innerHTML = res;
      });
      console.log(this.state.height);
      console.log(this.state.width);
      console.log(this.state.mines);
      console.log(this.state.className);
	  console.log(squares);
    });
  }
  
  render() {
    const {height, width, mines, gameStatus} = this.state
      return (
        <div className={this.state.className}>       
        {this.state && this.state.mines &&
          <Board height={this.state.height} width={this.state.width} mines ={this.state.mines} />
        }
        <p id = "Leaderboard"></p>
        </div>
      );
  }
 
}

ReactDOM.render(<Game />, document.getElementById("root"));




src/ServerApi.js
class ServerApi {

  /** 
   * API to get new game with difficulty(Easy, Medium, Hard) from backend server.
   * Returns an object with row number, column number, and mine field as a plain string.
   * 
   * Usage:
   *    let server = new ServerApi();
        server.getNewGame('Easy')
          .then(res => {
            console.log('rows - ' + res.row);
            console.log('cols - ' + res.col);
            console.log(res.field);
        });
   * 
   * Example return value:
   * {  row: 5, 
   *    col: 5,
   *    mines: 10, 
   *    field: "0100000010000110110110101"
   *    GameID: 5  }
  */
  async getNewGame(difficulty) {
    var response = await fetch('/api/newgame/' + difficulty);
    var body = await response.json();

    if (response.status !== 200) {
      throw Error(body.message) 
    }

    return body;
  }

  /** 
   * API to get leaderboard from backend server.
   * Returns the list of highscore data.
   * 
   * Example return value:
   * [ { Lid: 1, 
   *     Score: 543,
   *     Username: "Bob", 
   *     Difficulty: "medium" },
   *   ...,
   *   { Lid: 10, 
   *     Score: 950,
   *     Username: "Jack", 
   *     Difficulty: "medium" } ]
  */
  async getLeaderboard(difficulty){
    var response = await fetch('/api/leaderboard/' + difficulty);
    var body = await response.json();

    if (response.status !== 200) {
      throw Error(body.message) 
    }
    var result = "";
    for (let i = 0; i < body.list.length; i++) {
      if (body.list[i].Score == null) continue;
      let row = "Name: " + body.list[i].Username + " / Score: " + body.list[i].Score + "<br>";
      result += row;
    }
    return result;
  }

  /*
  Example return value:

  Game: 5 data deleted
  */
  async deleteUserData(GID){
    var response = await fetch('/api/deleteuserdata' + GID);
    var body = await response.json();

    if (response.status !== 200) {
      throw Error(body.message) 
    }

    return body;
  }

  /** 
   * API to check if the new score is going to be in top 10.
   * Returns true if the new score is in top 10, returns false otherwise.
  */
  async checkScore(difficulty, score) {
    var requestBody = {
      difficulty: difficulty,
      score: score
    };

    var config = {
      method: 'POST',
      headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody)
    }

    var response = await fetch('/api/score', config);
    var body = await response.json();

    if (response.status !== 200) {
      throw Error(body.message) 
    }

    return body.result;
  }

  /** 
   * API to insert new high score.
   * Returns the updated list of highscore data.
   * 
   * Example return value:
   * [ { Lid: 1, 
   *     Score: 543,
   *     Username: "Bob", 
   *     Difficulty: "medium" },
   *   { Lid: 2, 
   *     Score: 555,
   *     Username: "Jill", 
   *     Difficulty: "medium" }, ... ]
  */
  async insertNewScore(difficulty, score, userName) {
    var requestBody = {
      difficulty: difficulty,
      score: score,
      username: userName
    };

    var config = {
      method: 'POST',
      headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody)
    }

    var response = await fetch('/api/score', config);
    var body = await response.json();

    if (response.status !== 200) {
      throw Error(body.message) 
    }

    return body.list;
  }

}

export default ServerApi;

server/WebSweeper.java
import java.util.Random;
import java.util.Scanner;

public class WebSweeper {

	public static int rows;
	public static int cols;
	public static int mines;
	
	public static void main(String[] args) {
		
		int temp;
		
		Scanner in = new Scanner(System.in);
		
		do {
		System.out.println("Enter positive number of rows: ");
		temp = in.nextInt();
		}while(!(temp > 0));
		setRows(temp);
		
		do {
		System.out.println("Enter positive number of columns: ");
		temp = in.nextInt();
		}while(!(temp > 0));
		setCols(temp);
		
		do {
		System.out.println("Enter positive number of mines and less then rows * columns: ");
		temp = in.nextInt();
		}while(!(temp < rows*cols) && !(temp > 0));
		setMines(temp);
		
		
		int[][] field = new int[rows][cols];
		
		field = createField();
		
		printField(field);
		
		int row, col, num;
		
		while(true) {
			System.out.println("\nEnter row col to check adjacent mines, enter -1 to exit: ");
			row = in.nextInt();
			
			if(row == -1)
				break;
			
			col = in.nextInt();
			
			num = checkAdjacent(row, col, field);
			
			System.out.println("Number of adjacent mines is, -1 if mine: " + num);
		}
		
		in.close();
		
	}
	
	public static void setRows(int num) {
		
		rows = num;
		
	}
	
	public static void setCols(int num) {
		
		cols = num;
		
	}
	
	public static void setMines(int num) {
		
		mines = num;
		
	}
	
	public static int[][] createField() {
		
		int[] mineSpot = new int[mines];
		int[][] field = new int[rows][cols];
		int i, j;
		
		mineSpot = genMines();
		
		for(i = 0; i < rows; i++) {
			for(j = 0; j <cols; j++) {
				field[i][j] = 0;
			}
		}
		
		for(i = 0; i < mines; i ++) {
			
			int row = (mineSpot[i] / cols);
			int col = (mineSpot[i] % cols);
			field[row][col] = 1;
		}
		
		return field;
		
	}
	
	private static int[] genMines() {
		
		int[] mineSpot = new int[mines]; 
		int i, j, temp;
		
		for(i = 0; i < mines; i++) {
			mineSpot[i] = -1;
		}
		
		Random rand = new Random();
		
		for(i = 0; i < mines; i++) {
			
			temp = rand.nextInt(rows*cols);
			
			for(j = i; j >= 0; j--) {
				if(temp == mineSpot[j]) {
					temp = -2;
					break;
				}
			}
			
			if(temp == -2) {
				i--;
				continue;
			}
			
			mineSpot[i] = temp;
			
		}
		
		return mineSpot;
	}
	
	public static int checkAdjacent(int row, int col, int[][] field) {
		
		int num = 0;
		
		if(field[row][col] == 1)
			return -1;
		
		if(row - 1 >= 0)
			num += field[row -1][col];
		if(row - 1 >= 0 && col + 1 < cols)
			num += field[row -1][col +1];
		if(col +1 < cols)
			num += field[row][col +1];
		if(col +1 < cols && row +1 < rows)
			num += field[row +1][col +1];
		if(row +1 < rows)
			num += field[row +1][col];
		if(row +1 < rows && col -1 >= 0)
			num += field[row +1][col -1];
		if(col - 1 >= 0)
			num += field[row][col -1];
		if(col -1 >= 0 && row -1 >= 0)
			num += field[row -1][col -1];
		
		return num;
	}
	
	public static void printField(int[][] field) {
		
		int i, j;
		for(i = 0; i < rows; i++) {
			for(j = 0; j < cols; j ++) {
				System.out.print("" + field[i][j] + " ");
			}
			System.out.print("\n");
		}		
	}
}


src/components/Cell.js
import React from 'react';

// The class for the cells on the board
export default class Cell extends React.Component {

    getValue() { // Gets the value of what is in the cell
        if (!this.props.value.isRevealed) { // If the cell value is hidden
            return this.props.value.isFlagged ? "🏴" : null;
        }
        if (this.props.value.isMine) { // If the Cell is a mine
            return "💣";
        }
        if (this.props.value.neighbor == 0) { // If a neighboring cell does not have a value
            return null;
        }
        return this.props.value.neighbor; // Return the neighboring cells by default
    }

    render() {
        let className = "cell" + (this.props.value.isRevealed ? "" : " hidden") + (this.props.value.isMine ? " is-mine" : "") + (this.props.value.isFlagged ? " is-flag" : "");

        return (
            <div ref = "cell" onClick = {this.props.onClick} className = {className} onContextMenu = {this.props.cMenu}>
                {this.getValue()}
            </div>
        );
    }
}

src/components/Board.js
import React from 'react';
import Cell from './Cell'; // import the cell component
import ServerApi from './ServerApi'; // import the Server Api

export default class Board extends React.Component {
    state = {
        boardData: this.boardDataInit(this.props.width, this.props.height, this.props.mines),
        gameWin: false,
        numMines: this.props.mines
    };

    render() {
        return (
            <div className="board">
                <div className="game-info">
                    <span className="info">mines: {this.state.numMines}</span><br />
                    <span className="info">{this.state.gameWin ? "You Win!" : ""}</span>
                </div>
                {
                    this.renderBoard(this.state.boardData)
                }
            </div>
        );
    }

    /* HELPER METHODS */

    // Get the flags array
    getFlagsArr(flagsData) {

    }

    // Get the mines array
    getMinesArr(minesData) {

    }

    // Get the initial board data
    boardDataInit(width, height, mines) {

    }

    // Place the mines on the board
    placeMines(arrData, width, height, mines) {

    }

    // Get the number of neighboring mines
    getNumNeighboring() {

    }

    /* EVENT HANDLING */
    
}

server/server.js
const express = require('express');
const bodyParser = require('body-parser');
const mysql = require('mysql');
const app = express();

const PORT = 5000;

var rows;
var cols;
var mines;
var diff;
var gameType;
var GID;

// mySQL connection info
const con = mysql.createConnection({
    host: "cs319-121.misc.iastate.edu",
    user: "admin",
    password: "WebSweeper2019!",
    database: "WebSweeper"
});

app.use(bodyParser.json());

app.get('/', (req, res) => {
    res.redirect('/api/connect');
});

/*
* GET /api/connect
*/
app.get('/api/connect', (req, res) => {
    res.status(200).send({ message: 'Hello from backend server' });
});

/*
* GET /api/newgame/[DIFFICULTY]
*    - example request: GET /api/newgame/Easy
*/
app.get('/api/newgame/:difficulty', (req, res) => {
    var status = 400;
    var message = {};
 
    console.log('difficulty: ' + req.params.difficulty);

    diff = req.params.difficulty;

    if(diff == "easy"){
        rows = 5;
        cols = 5;
        mines = 5;
        gameType = 'gameEasy';
    }else if(diff == "medium"){
        rows = 8;
        cols = 8;
        mines = 10;
        gameType = 'game';
    }else if(diff == "hard"){
        rows = 15;
        cols = 15;
        mines = 50;
        gameType = 'gameHard';
    }else if(diff == "extreme"){
        rows = 15;
        cols = 15;
        mines = 150;
    }else if(diff == "psycho"){
        rows = 15;
        cols = 15;
        mines = 200;
    }
    
    var field = createField();
    var stringField = fieldStringify(field);
    var viewField = viewFieldGen();

    var sql = "insert into Game (Field, roe, col, mines, Difficulty, viewField) " 
    sql += "Values(\'" + stringField +"\', "+ rows +", " + cols + ", "+ mines+", \'" + diff +"\', \'"+ viewField + "\')";

    con.query(sql, function (err, result){
        if(err) throw err;
        status = 200;
        GID = result.insertId;
        message = {gameType: gameType, mine: mines, row: rows, col: cols, field: stringField, GameID: GID};
        res.status(status).send(message);
        console.log("New Game Inserted. Game ID: " + GID);
    })

    
});

/*
* GET /api/leaderboard/[DIFFICULTY]
*    - example request: GET /api/leaderboard/easy
*/
app.get('/api/leaderboard/:difficulty', (req, res) => {
    var sql = "SELECT * FROM Leaderboard " +
        "WHERE Difficulty = \'" + req.params.difficulty + "\' ORDER BY Score LIMIT 10"; 

    con.query(sql, function (err, result){
        if(err) throw err;
        res.status(200).send({ list: result });
    });
});

/*
* GET /api/leaderboard
*/
app.get('/api/viewfield/:gameID', (req, res) => {

    gameID = req.params.gameID;

    var sql = "select viewField from Game where GID = " + gameID;
    
    con.query(sql, function (err, result){
        if(err) throw err;
        status = 200;
        res.status(status).send(result);
        console.log("ViewField for GID: " + gameID + " is: " + result[0].viewField);
    })

});

app.get('/api/deletedata', (req, res) =>{

    var sql = "DELETE FROM Game;"
    con.query(sql, function (err, result){
        if(err) throw err;
        console.log("Game Table Data Deleted");
    })
    sql = "DELETE FROM Leaderboard;"
    con.query(sql, function (err, result){
        if(err) throw err;
        console.log("Leaderboard Table Data Deleted");
    })

    status = 200;
    res.status(status).send("Data Deleted");
})

app.get('/api/deleteuserdata/:gameID', (req, res) => {

    gameID = req.params.gameID;

    var sql = "DELETE FROM Game WHERE GID = " + gameID;

    con.query(sql, function (err, result){
        if(err) throw err;
        console.log("Game: " + gameID + " Data Deleted");
    })

    status = 200;
    res.status(status).send("Game: " + gameID + " Data Deleted");
})

/*
* POST /api/score
*      {  "difficulty" : [DIFFICULTY],
*         "score" : [SCORE],
*         "username" : [USERNAME (OPTIONAL)]  }
*
*    - example request payload (usecase 1 - check if the new score is in top10): 
*          {  "difficulty" : "easy",
*             "score" : 433  }
*    - example request payload (usecase 2 - insert new score): 
*          {  "difficulty" : "easy",
*             "score" : 433,
*             "username" : "Bob"  }
*/
app.post('/api/score', (req, res) => {
    var payload = req.body;
    console.log(payload);

    var sqlSelect = "SELECT * FROM Leaderboard " +
        "WHERE Difficulty = \'" + payload.difficulty + "\' ORDER BY Score LIMIT 10"; 

    con.query(sqlSelect, function (err, result){
        if(err) throw err;

        // check if the new score is in top 10
        var response = false;
        var lId = -1;
        for (var i = 0; i < result.length; i++) {
            if (result[i].Score == null || result[i].Score > payload.score) {
                response = true;
                lId = result[i].LID;
                break;
            }
        }

        if (payload.username == null || payload.username == undefined || payload.username === '' || lId == -1) {
            res.status(200).send({ result: response, lId: lId });
            return;
        }
        
        // update leaderboard (if the payload contains username)
        var sqlUpdate = "insert into Leaderboard (score, Username, Difficulty) values (" + payload.score + 
                 ", \'" + payload.username + "\', \'" + payload.difficulty + "\')";
        con.query(sqlUpdate, function (err, result) {
            if(err) throw err;

            // get updated leaderboard data
            con.query(sqlSelect, function (err, result){
                if(err) throw err;
                res.status(200).send({ result: response, lId: lId, list: result });
            });
        });

    });
});

app.listen(PORT, () => console.log('Server listening on port ' + PORT));

// connect database
con.connect((err) => {
    if (err) {
        throw err;
    }

    console.log('Database connected');
    con.query('SHOW TABLES', (err, result) => {
        if (err) throw err;
        console.log(result);
    });
})

function createField(){
    var mineSpot = [];
    var field = [];
    var i, j;

    for(i = 0; i < rows; i++){
        field[i] = new Array(cols);
    }

    mineSpot = genMines();

    for(i = 0; i < rows; i++){
        for(j = 0; j < cols; j++){
            field[i][j] = 0;
        }
    }

    for(i = 0; i < mines; i++){
        var row = Math.floor((mineSpot[i] / cols));
        var col = Math.floor((mineSpot[i] % cols));
        field[row][col] = 1;
    }

    return field;
}

function genMines(){
    var mineSpot = [];
    var i, j, temp;

    for(i = 0; i < mines; i++){
        mineSpot[i] = -1;
    }

    for(i = 0; i < mines; i++){
        temp = Math.floor(Math.random() * rows * cols);

        for(j = i; j >= 0; j--){
            if(temp == mineSpot[j]){
                temp = -2;
                break;
            }
        }

        if(temp == -2){
            i--;
            continue;
        }

        mineSpot[i] = temp;
    }
    return mineSpot;
}

function fieldStringify(field){
    var i, j;
    var stringField = "";

    for(i = 0; i < rows; i++){
        for(j = 0; j < cols; j++){
            stringField += "" + field[i][j];
        }
    }

    return stringField;
}

function viewFieldGen(){
    var i, j;
    var vField = "";

    for(i = 0; i < rows; i++){
        for(j = 0; j < cols; j++){
            vField += "0";
        }
    }

    return vField;
}

module.exports = app;   // for unittest